File: 0-99/014-longest-common-prefix.cpp
#include <iostream>
#include <cstdio>
#include <cmath>
#include <cstdlib>
#include <cstring>
#include <string>
#include <queue>
#include <climits>
using namespace std;

class Solution
{
  public:
    string GetSubStr(string s, int index)
    {
        if (index <= 0)
            return "";
        string rr(s, 0, index);
        return rr;
    }
    string longestCommonPrefix(vector<string> &strs)
    {
        int size = strs.size(), i, index = 0;
        if (size <= 0)
            return "";
        char c;
        while (true)
        {
            if (strs[0].length() <= index)
            {
                return GetSubStr(strs[0], index);
            }
            c = strs[0][index];
            for (i = 1; i < size; i++)
            {
                if (strs[i].length() <= index || strs[i][index] != c)
                {
                    return GetSubStr(strs[i], index);
                }
            }
            index++;
        }
    }
};

File: 0-99/014-longest-common-prefix.py
# -*- coding:utf-8 -*-
'''
@Author: lamborghini1993
@Date: 2020-06-15 20:14:20
@Description: https://leetcode-cn.com/problems/longest-common-prefix/
'''


from typing import List


class Solution:
    def same(self, index, strs):
        if len(strs[0]) <= index:
            return False
        c = strs[0][index]
        for string in strs:
            if len(string) <= index or string[index] != c:
                return False
        return True

    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""
        index = 0
        while True:
            if self.same(index, strs):
                index += 1
            else:
                return strs[0][:index]
        return ""


func = Solution().longestCommonPrefix
print(func(["flower", "flow", "flight"]))
print(func(["flower", "flow", ""]))
print(func([]))
print(func(["dog", "racecar", "car"]))
